### DISCUSSION TITLE
RFC: Printing and Capability Security

### DISCUSSION THREAD

**Discussion by eudoxia0** [austral/austral#392] [Ideas]
# Summary

How do we integrate capability-based security for access to the terminal, without inconveniencing new users too much?

# Problem

Printing is an effect. It should probably require a capability to do. This can also help with ensuring the terminal is used in a single-threaded way.

But that means that a "Hello, world!"-like program will require some capability boilerplate. My concern is that potential users of the language read the tutorial, see that the "Hello, world!" program is ten lines of obscure API calls, and walk away instantly.

As I said in #341, there are two options:

- **Capability-secure terminal IO:** most people will open the tutorial, see that "Hello, world!" is like seven lines of code for the capability setup (and the more granular the capabilities, the longer the code gets), and close the tab right then and there. It is asking a lot of people to go "please ignore all this giant seeming boilerplate".
- **No capability-secure terminal IO:** then people can complain that it's not Truly Secure.

# Solution

So here's the Solomonic compromise:

1. The standard library will capability-secure terminal IO.
2. When compiling in `debug` mode (the default), there will be a module called something like `Austral.Debug` whose contents are implicitly imported, including methods like `debugPrint` or `debugPrintLn`.

So then, the "Hello, world!" program goes from this:

```austral
module body Hello is
    function main(): ExitCode is
        printLn("Hello, world!");
        return ExitSuccess();
    end;
end module body.
```

To something like this:

```austral
module body Hello is
    function main(): ExitCode is
        debugPrintLn("Hello, world!");
        return ExitSuccess();
    end;
end module body.
```

Which is not too terribly demanding.

The "proper", capability-secure, production-ready version would be something like:

```austral
-- Imports elided.

module body HelloWorld is
    function main(root: RootCapability): ExitCode is
        let termcap: TerminalCapability := acquireTermCap(&!root);
        printLn(&termcap, "Hello, world!");
        surrenderTermCap(termcap);
        surrenderRoot(root);
        return ExitSuccess();
    end;
end module body.
```

# Questions

1. Syntax, names, etc. `debugPrint` and `debugPrintLn` are rather long. Perhaps `echo`.
2. Is it worth throwing people in the deep end of capabilities from the start?

# See Also

- #336 
- #341

**Comment by ckp95** [comment]
Some ideas to make the capability-secure hello-world more palatable:

you could let main ask for granular capabilities in the parameter list, instead of the root capability:

```
module body HelloWorld is
    function main(stdout_cap: StdoutCapability): ExitCode is
        printLn(&stdout_cap, "Hello, world!");
        surrenderStdoutCap(stdout_cap);
        return ExitSuccess();
    end;
end module body.
```

(then the capabilities given to `main` could be constrained by compiler options)

can the surrender functions be generic?

```
typeclass Surrenderable(T: Linear) is
    method surrender(t: T): Unit;
end;
instance Surrenderable(RootCapability);
instance Surrenderable(TerminalCapability);
...
```

so you wouldn't have the redundancy of `surrenderFoo(foo);`, it would be:

```
surrender(termcap);
surrender(root);
```

----

>Syntax, names, etc. debugPrint and debugPrintLn are rather long. Perhaps echo.

`echo` and `errecho`? or you could "gently encourage" devs not to rely on debug mode by making the `echo` only print to stderr, not stdout.

>Is it worth throwing people in the deep end of capabilities from the start?

It may be enough to show the two examples and the reason for the tradeoffs on the introductory page, and that it's okay to rely on the `--debug` version while going through tutorials. Capabilities are a headline feature of Austral, and a high chance of why beginners are reading about it in the first place, so it shouldn't be completely obscured.

----

EDIT I had this in the first part of my comment but I'm less sure about it the more I think about it bc the syntax seems too special-case, and it's not clear how errors would be handled:

<details>
<summary>click to expand</summary>
acquiring a linear value, passing it into precisely one function, then immediately destroying it, may become a common pattern. there could be syntax sugar:

```
# provisionally use "~>" to show the idea
some_function(source_value~>acquire_func~>discard_func, param2, param3);
```
expanding to
```
let temp: TempType := acquire_func(source_value); # does this count as type inference?
some_function(temp, param2, param3);
discard_func(temp);
```

or a two-argument form `source_value~>discard_func` for where you don't need to acquire it first.

so you could write
```
printLn(&!root~>acquireStdout~>surrender, "Hello, world!");
```
or
```
printLn(stdout_cap~>surrender, "Hello, world!");
```

combining the above ideas, Hello World could be:

```
module body HelloWorld is
    function main(stdout_cap: StdoutCapability): ExitCode is
        printLn(&stdout_cap~>surrender, "Hello, world!");
        return ExitSuccess();
    end;
end module body.
```
</details>

**Comment by andyferris** [comment]
@eudoxia0 - first thanks for introducing Austral! I hope it's OK I jump in here with an opinion.

> My concern is that potential users of the language read the tutorial, see that the "Hello, world!" program is ten lines of obscure API calls, and walk away instantly.

To be honest, I experienced the exact opposite effect. I was nodding along in agreement to "Introducing Austral" discussing the rational behind the capability system and the root capability, and then was very surprised that _hello world_ has a `main()` function without a root capability, and there was seemingly unfettered access to stderr/stdout (and perhaps also stdin?). How do I ensure my imports aren't reading and writing to file descriptors? To use the logging library example, if I give my logger only one capability - to access to a single directory - I assume it would be impossible for it to print logs stdout.

What I expected to see was something like:

```
module body HelloWorld is
    function main(root: RootCapability): Unit
        printLn(&! root.stdout, "Hello, world!");
        ExitSuccess(root);
    end;
end module body.
```

(Sorry I'm not 100% sure whether that is the right syntax for a mutable borrow of a linear field of an owned linear record... maybe it's `root.&!stdout`?)

In my opinion this is concise enough, ensures the stdout output has no concurrency issues, etc.

You'll also note another "capability" I've added is the capability to terminate the program. I've been bitten by imported code that called `exit()` instead of bubbling up errors, and it can be quite impossilbe to work around (e.g. even a thread worker could terminate the process). Similarly, I have heard some of the rust folks are looking into "provably panic-free code". In the above exiting is _the_ effect that consumes `root` (but it could probably be split into it's own more granular capability that can be more safely passed around and/or be conditionally invoked).

**Reply by crabmusket** [reply]
I like this! Now the "hello world" example becomes quite distinctive: it shows off the interesting features of the language while being reasonably succinct.

**Comment by crabmusket** [comment]
I think the main concern I have with printing and capabilities is not the hello-world case (on which I agree with what Andy said), but the "print debugging" case. This is still something I do with embarrassing regularity, even in environments with decent debugging tools. When debugging pure code, it'd be extremely tedious to have to invade the entire module hierarchy with a terminal capability, then remove it again.

> When compiling in debug mode (the default), there will be a module called something like Austral.Debug whose contents are implicitly imported, including methods like debugPrint or debugPrintLn.

What would happen in non-debug builds? Does `debugPrint` become a compile error? Or could it become a no-op? The former seems more in-line with Austral's strictness*. On the other hand, it may be a benefit to have libraries which ship debug diagnostics so they can be traced by their users in debug builds. (And then that gets me started thinking about enabling debug traces per-module, which is starting to get way off topic.)

*Sort of like Go's insistence that no unused variables or imports may compile.

<details>

<summary>A digression about debug/tracing and linear types</summary>

Thinking about "debug statements", I started to think of tracing. Then I realised linear types are pretty good for that, I think?

```
import Austral.Debug (Scope, enter, exit);

function something(): Unit is
  scope: Scope := enter("some function");
  doThings();
  exit(scope);
  return nil;
end;
```

As with any other resource, the type system ensures you do your tracing properly.

</details>

**Comment by eudoxia0** [comment]
>Does debugPrint become a compile error?

Yes. I'm not sure if making it a no-op can be done without causing surprising edge cases (e.g. `debugPrint(f(x))` where `f(x)` consumes a linear variable `x`).

>On the other hand, it may be a benefit to have libraries which ship debug diagnostics so they can be traced by their users in debug builds.

No, I don't think this is good. This should be done by a logging framework you can turn on/off.
